### DISCUSSION TITLE
Can we expect Parallel Consumer to work with Kafka 4.0.0 client (or 3.9.1)?

### DISCUSSION THREAD

**Discussion by jvissers** [confluentinc/parallel-consumer#848] [General]
My primary reasons for asking this, is because of Java 24 uptake. 

One of the things that will be in that version is the permanent removal of the security manager. We are currently on Java 23, and we have to use the flag `-Djava.security.manager=allow` to be able to run our service. That option will no longer be available in Java 24, and this requires us to migrate to [Kafka 4.0.0](https://cwiki.apache.org/confluence/display/KAFKA/Release+Plan+4.0.0), once that hits. Curious whether we are expecting this to be a drop-in replacement, without disruptions to the Parallel Consumer library.

**Comment by jvissers** [comment]
@astubbs - gentle reminder for the above open question?

**Comment by bernardmo** [comment]
I'll just add a note for java 24 setup - in some cases it's possible to force `kafka-clients` lib dependency to `4.0.0` and set the `ignoreReflectiveAccessExceptionsForAutoCommitDisabledCheck` to `true`. Not sure what are the side-effects of that kind of setup, but the comment states:
```
The purpose of the flag is to be a last resort / temporary work-around for changes introduced in newer Kafka Clients that break reflective access and for using wrapped, custom or extended KafkaConsumer classes that fail reflection checks - setting the flag to true will ignore reflection access exceptions during this check.
Note: that library will still try to access auto commit field on the consumer object and if it is accessible and not disabled - the Parallel Consumer will shut down.
```

**Reply by jvissers** [reply]
Thanks @bernardmo - at the moment Kafka 3.9.1 is cooking over at Kafka (currently RC1). This version will have a backport of the security manager shim that is needed to run this on Java 24. Do you anticipate that parallel consumer will run unchanged on that specific version?

**Reply by bernardmo** [reply]
Hi @jvissers.

> This version will have a backport of the security manager shim that is needed to run this on Java 24. 

That's great to know. 

> Do you anticipate that parallel consumer will run unchanged on that specific version?

At the moment I have no idea. My comment was placed just to note that the current version of parallel consumer can run on java 24 with kafka clients 4.0.0 (was forced to upgrade the version on one of the projects that I work on due to virtual thread blocking issue). For now I didn't noticed any issues, but again it's a limited use case.

**Comment by astubbs** [comment]
@jvissers, @bernardmo - yes, to both, with one qualification on 4.0.

**kafka-clients 3.9.1 and later: yes, released.** [v0.6.0.0](https://github.com/astubbs/parallel-consumer/releases/tag/v0.6.0.0) of the maintained fork ships on kafka-clients 3.9.2 and runs unchanged on Java 24. The Java 24 breakage was never in Parallel Consumer: it was kafka-clients' SASL callback calling `Subject.getSubject`, which JDK 24 removed the escape hatch for, and 3.9.1 fixed it (KAFKA-19024). The library has no reference to the security manager, `AccessController` or JDK internals, and is Java 8 bytecode; an end-to-end run over a mock consumer completed identically on JDK 17 and JDK 26. Your #862 is mirrored as astubbs/parallel-consumer#181 and closed on that basis. What is not claimed: a real SASL client on a modern JDK in CI - that lane is astubbs/parallel-consumer#128, open.

**kafka-clients 4.0.0: yes, and properly soon.** The fork will shortly move its baseline to Kafka 4 so that the 4.x client is simply supported, built and tested against, with nothing to work around - that is astubbs/parallel-consumer#53, open. Until it lands, @bernardmo's setup works, and here is why. The APIs 4.x removed - `sendOffsetsToTransaction(Map, String)`, `new ConsumerGroupMetadata(String)`, the `MockConsumer(OffsetResetStrategy)` constructor - sit on a deprecated overload PC's own code does not call and on the test mock consumer; the transactional commit path calls the `ConsumerGroupMetadata` overload, which 4.x keeps. So with the 4.0 client on the classpath every production path resolves, and `ignoreReflectiveAccessExceptionsForAutoCommitDisabledCheck` is exactly what it says - it bypasses the reflective auto-commit check that 4.0's consumer internals broke, nothing more. What you do not get yet is a build or a test run against 4.x: the fork compiles against 3.9 today, and the baseline move (Java 11 minimum, the removed APIs migrated, the experimental Kafka 4.x CI job switched back on) is what astubbs/parallel-consumer#53 does. Reports from a 4.0 deployment are welcome on that issue; @bernardmo's is the first.

The coordinates changed with the fork - on Maven Central it is now:

```xml
<dependency>
    <groupId>bz.stub.parallelconsumer</groupId>
    <artifactId>parallel-consumer-core</artifactId>
    <version>0.6.0.0</version>
</dependency>
```

(`parallel-consumer-vertx`, `parallel-consumer-reactor` and `parallel-consumer-mutiny` likewise.) The Java package is `bz.stub.parallelconsumer` too, so imports change with it - one command rewrites them (GNU `sed`; on macOS use `sed -i ''`):

```bash
find . -name '*.java' -exec sed -i 's/io\.confluent\.parallelconsumer/bz.stub.parallelconsumer/g' {} +
```
